goal_src/engine/gfx/shadow/shadow.gc

```diff
@@ -5,3 +5,339 @@
+      (set! (-> sv-144 x) (-> sv-96 z))
+      (set! (-> sv-144 y) 0.0)
+      (set! (-> sv-144 z) (- (-> sv-96 x)))
+      (vector-normalize! sv-144 1.0)
+      (set! sv-112 quaternion-vector-angle!)
+      (set! sv-128 s0-0)
+      (let ((a2-1 (acos (-> sv-96 y))))
+       (sv-112 (the-as quaternion sv-128) sv-144 a2-1)
+       )
+      (let ((v1-9 s1-0))
+       (cond
+        ((< (-> s0-0 w) 0.0)
+         (.lvf vf1 (&-> v1-9 vel-sxvel quad))
+         (.lvf vf2 (&-> s0-0 quad))
+         (.sub.vf vf1 vf0 vf2 :mask #b111)
+         (.svf (&-> v1-9 vel-sxvel quad) vf1)
+         (.mov v1-10 vf1)
+         )
+        (else
+         (.lvf vf1 (&-> v1-9 vel-sxvel quad))
+         (.lvf vf2 (&-> s0-0 quad))
+         (.add.vf vf1 vf0 vf2 :mask #b111)
+         (.svf (&-> v1-9 vel-sxvel quad) vf1)
+         (.mov v1-11 vf1)
+         )
+        )
+       )
+      (draw-shadow arg0 arg1 (-> s1-0 vel-sxvel) (the-as float s4-0) s3-0 s2-0)
+      )
+     )
+    )
+   0
+   (none)
+   )
+  )
+
+;; definition for function find-ground-and-draw-shadow
+;; INFO: Return type mismatch int vs none.
+;; Used lq/sq
+(defun
+  find-ground-and-draw-shadow
+  ((arg0 vector)
+   (arg1 vector)
+   (arg2 float)
+   (arg3 uint)
+   (arg4 process)
+   (arg5 float)
+   (arg6 float)
+   )
+  (let ((s2-0 (new 'stack-no-clear 'vector)))
+   (set! (-> s2-0 quad) (-> arg0 quad))
+   (new 'stack-no-clear 'vector)
+   (+! (-> s2-0 y) arg5)
+   (let ((s4-0 (new 'stack-no-clear 'bone)))
+    (cond
+     ((>=
+       (dummy-12 *collide-cache* s2-0 arg6 arg3 arg4 s4-0 (the-as uint 1))
+       0.0
+       )
+      (if (!= arg2 0.0)
+       (compute-and-draw-shadow
+        s2-0
+        (-> s4-0 position)
+        (-> s4-0 scale)
+        (the-as vector arg2)
+        arg6
+        (the-as float 0)
+        )
+       )
+      (if (and arg1 (!= (shr (shl (-> s4-0 cache bone-matrix) 52) 58) 3))
+       (set! (-> arg1 quad) (-> s4-0 position quad))
+       )
+      )
+     (else
+      (if arg1
+       (vector+float*!
+        arg1
+        arg0
+        (-> *standard-dynamics* gravity-normal)
+        (- arg6)
+        )
+       )
+      )
+     )
+    )
+   )
+  0
+  (none)
+  )
+
+;; definition for function do-target-shadow
+;; INFO: Return type mismatch int vs none.
+;; Used lq/sq
+(defbehavior do-target-shadow target ()
+  (if
+   (and
+    (logtest? (-> self control status) 1)
+    (!= (-> self control unknown-surface mode) 'swim)
+    (!= (-> self control unknown-surface mode) 'dive)
+    (!= (-> self next-state name) 'target-flop)
+    (zero? (logand (-> self draw status) 38))
+    )
+   (set! (-> self control shadow-pos quad) (-> self control trans quad))
+   (find-ground-and-draw-shadow
+    (-> self control trans)
+    (-> self control shadow-pos)
+    0.0
+    (-> (the-as collide-shape-prim (-> self control root-prim)) collide-with)
+    self
+    0.0
+    81920.0
+    )
+   )
+  0
+  (none)
+  )
+
+;; failed to figure out what this is:
+(set!
+  (-> *part-id-table* 362)
```

Comment:
It looks like there's a crash when this code is loaded. Maybe we haven't decompiled the file that creates `*part-id-table*` yet? You could try to find where it's defined and just add that part, or just comment this out for now (and maybe leave a comment near the top of the file so we don't forget this)

Reply:
ah yeah we havn't. that's either `sparticle` or `sparticle-launcher`